packages/mobile/src/components/audio/AudioPlayer.tsx

```diff
@@ -626,44 +608,76 @@ export const AudioPlayer = () => {
         artwork: imageUrl,
         duration: isPreview ? getTrackPreviewDuration(track) : track.duration
       }
-    })
+    }
+
+    // Enqueue tracks using 'middle-out' to ensure user can ready skip forward or backwards
+    const enqueueTracks = async (
+      queuableTracks: QueueableTrack[],
+      queueIndex = 0
+    ) => {
+      let currentPivot = 1
+      while (
+        queueIndex - currentPivot > 0 ||
+        queueIndex + currentPivot < queueTracks.length
+      ) {
+        if (abortEnqueueControllerRef.current.signal.aborted) {
+          return
+        }
+
+        const nextTrack = queuableTracks[queueIndex + currentPivot]
+        if (nextTrack) {
+          await TrackPlayer.add(await makeTrackData(nextTrack))
+        }
+
+        const previousTrack = queuableTracks[queueIndex - currentPivot]
+        if (previousTrack) {
+          await TrackPlayer.add(await makeTrackData(previousTrack), 0)
+        }
+        currentPivot++
+      }
+    }
 
     if (isQueueAppend) {
-      await TrackPlayer.add(newTrackData)
+      enqueueTracksJobRef.current = enqueueTracks(newQueueTracks)
+      await enqueueTracksJobRef.current
+      enqueueTracksJobRef.current = undefined
     } else {
-      // New queue, reset before adding new tracks
-      // NOTE: Should only happen when the user selects a new lineup so reset should never be called in the background and cause an error
       await TrackPlayer.reset()
-      await TrackPlayer.add(newTrackData)
-      if (queueIndex < newQueueTracks.length) {
-        await TrackPlayer.skip(queueIndex)
+
+      const firstTrack = newQueueTracks[queueIndex]
+      if (!firstTrack) return
+      await TrackPlayer.add(await makeTrackData(firstTrack))
+
+      if (playing) {
+        await TrackPlayer.play()
       }
-    }
 
-    if (playing) await TrackPlayer.play()
-    updatingQueueRef.current = false
+      enqueueTracksJobRef.current = enqueueTracks(newQueueTracks, queueIndex)
+      await enqueueTracksJobRef.current
+      enqueueTracksJobRef.current = undefined
+    }
   }, [
-    isNotReachable,
-    isOfflineModeEnabled,
-    offlineAvailabilityByTrackId,
-    playing,
     queueIndex,
-    queueTrackOwnersMap,
     queueTrackUids,
-    queueTracks,
     didOfflineToggleChange,
+    queueTracks,
+    queueTrackOwnersMap,
+    isOfflineModeEnabled,
+    offlineAvailabilityByTrackId,
     isCollectionMarkedForDownload,
-    handleGatedQueryParams,
-    isReachable,
-    storageNodeSelector
+    isNotReachable,
+    storageNodeSelector,
+    nftAccessSignatureMap,
+    playing
   ])
 
   const handleQueueIdxChange = useCallback(async () => {
     const playerIdx = await TrackPlayer.getActiveTrackIndex()
     const queue = await TrackPlayer.getQueue()
 
+    await enqueueTracksJobRef.current
```

Comment:
technically, wont this await the entire enqueue? it should only technically need to await the specific index right?

instead of a global ref, we could keep track of a metadata array for each index in the queue right?

const trackMetadata = useRef<(Promise<void> | metadata)[]>([])

and then append a job to that (at the right index) every time we enqueue and then when playing back an individual track we can just `await trackMetadata.current[idx]`

Reply:
Oh yeah I love this, so we don't have to spend any time building the queue. Two thoughts. 

1. This would make every new track take longer to play right? We technically do this on web though so probably chill?
2. I worry about when the app is in the background. Track-player is designed to keep working when app is backgrounded because it runs a worklet or whatever it's called. With this approach we are relying on the app running computations while in background which may get dropped?

Reply:
1.  I don't think so. You could still try to evaluate them as soon as they're added to the queue, just not block on it. Just make sure to collect the results individually per id

2. I think these computations are fast enough, it won't be an issue. iOS and android give you at least a couple minutes before they kill things, so if on any playback, we start the process of computing all the metadata we will need, it should fully resolve before things go dark

Reply:
great, i think i assumed that the array was delay computation approach rather than just tracking each one separately, which im down to switch to. fwiw the reason it's all one promise, is that we want to cancel the entire computation if we switch lineups, but we can easily accomplish that by promise,all'ing the array

Reply:
all of this is pretty rare since it finishes up fast, maybe it's worth us discussing live to go through all the requirements.